deeppavlov/core/models/lr_scheduled_model.py

```diff
@@ -0,0 +1,345 @@
+        self.start_val, self.end_val = start_val, end_val
+        self.iters = 0
+        if self.end_val is None and not (self.dec_type in [1, 4]):
+            self.end_val = 0
+        if self.dec_type == DecayType.ONECYCLE:
+            self.cycle_nb = math.ceil(self.nb / 2)
+            self.div = 1.0 if not self.start_val else self.end_val / self.start_val
+        if self.dec_type == DecayType.TRAPEZOID:
+            self.div = 1.0 if not self.start_val else self.end_val / self.start_val
+
+    def __str__(self):
+        return f"DecayScheduler(start_val={self.start_val}, end_val={self.end_val}"\
+            f", dec_type={self.dec_type.name}, num_it={self.nb}, extra={self.extra})"
+
+    def next_val(self):
+        self.iters = min(self.iters + 1, self.nb)
+        # print(f"iters = {self.iters}/{self.nb}")
+        if self.dec_type == DecayType.NO:
+            return self.start_val
+        elif self.dec_type == DecayType.LINEAR:
+            pct = self.iters / self.nb
+            return self.start_val + pct * (self.end_val - self.start_val)
+        elif self.dec_type == DecayType.COSINE:
+            cos_out = math.cos(math.pi * self.iters / self.nb) + 1
+            return self.end_val + (self.start_val - self.end_val) / 2 * cos_out
+        elif self.dec_type == DecayType.EXPONENTIAL:
+            ratio = self.end_val / self.start_val
+            return self.start_val * (ratio ** (self.iters / self.nb))
+        elif self.dec_type == DecayType.POLYNOMIAL:
+            delta_val = self.start_val - self.end_val
+            return self.end_val + delta_val * (1 - self.iters / self.nb) ** self.extra
+        elif self.dec_type == DecayType.ONECYCLE:
+            if self.iters > self.cycle_nb:
+                # decaying from end_val to start_val for cycle_nb steps
+                pct = 1 - (self.iters - self.cycle_nb) / self.cycle_nb
+                return self.start_val * (1 + pct * (self.div - 1))
+            else:
+                # raising from start_val to end_val for cycle_nb steps
+                pct = self.iters / self.cycle_nb
+                return self.start_val * (1 + pct * (self.div - 1))
+        elif self.dec_type == DecayType.TRAPEZOID:
+            if self.iters > 0.6 * self.nb:
+                # decaying from end_val to start_val for 4/10 * nb steps
+                pct = 2.5 * (self.nb - self.iters) / self.nb
+                return self.start_val * (1 + pct * (self.div - 1))
+            elif self.iters > 0.1 * self.nb:
+                # constant end_val
+                return self.end_val
+            else:
+                # raising from start_val to end_val for 1/10 * nb steps
+                pct = 10.0 * self.iters / self.nb
+                return self.start_val * (1 + pct * (self.div - 1))
+
+
+DType = Union[str, DecayType]
+
+
+class LRScheduledModel:
+    """
+    Abstract model enhanced with optimizer, learning rate and momentum
+    management and search.
+    """
```

Comment:
Constructor arguments list and their descriptions in this docstring would be great

Reply:
done